File: validator/src/validator.py
import sys
import os
import unittest
import subprocess
from subprocess import CalledProcessError, Popen

# Helper Class that has the job of running all the tests for the
# ValidatorListener class
class Validator:
    def __init__(self, test_path):
        self.path = test_path

    # Runs all the unit tests for a job
    def run_tests(self):
        result_key = ""
        try:
            print("installing dependencies...")
            result = Popen(["npm install"], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, cwd=self.path)
            result.wait()
            print("running tests...")
            result = Popen(["truffle", "test"], stdout=subprocess.PIPE, stderr=subprocess.PIPE, cwd=self.path)
            output = result.stdout.read().decode('utf-8')
            if("passing" in output.split('\n')[-3]):
                f = open(self.path + "secret_key","r")
                result_key = f.read()
                f.close()
        except Exception as e:
            result_key = e.output.decode('utf-8')
        return result_key

File: validator/src/validator.c
#include <openssl/pem.h>
#include <openssl/ssl.h>
#include <openssl/rsa.h>
#include <openssl/evp.h>
#include <openssl/bio.h>
#include <openssl/err.h>

#include "encrypt.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Helper function that concatenates 3 strings
char* concat(const char *s1, const char *s2, const char *s3) {
	char *result = malloc(strlen(s1) + strlen(s2) + strlen(s3) + 1);
	strcpy(result, s1);
	strcat(result, s2);
	strcat(result, s3);
	return result;
}

int main( int argc, char *argv[] ) {
	// Retrieve Command Line Arguments
	if(argc < 2) {
		printf("Usage: validator.o [PATH TO FILE]\n");
		exit(1);
	}
	char* path = argv[1];

	// Create the shell command to run tests
	char* cmd = concat("cd ", path, " && npm install && truffle test");
	printf("%s\n",cmd);

	// Initialize the shell pipe and buffer
	FILE *in;
	extern FILE *popen();
	char buff[512];

	// Run the shell command 
	if(!(in = popen(cmd, "r"))) {
		exit(1);
	}

	// Retrieve the output from the shell command
	while(fgets(buff, sizeof(buff), in)!=NULL) {
		printf("%s", buff);
	}

	// Close the shell pipe
	pclose(in);

	return 0;
}